README
Please check out the /docs directory for further information.

CHANGELOG.md
# Change Log
Only partial changelog, [commit history](https://framagit.org/Shnoulle/LimeSurvey/commits/2.06_SondagesPro) show all changelog.

## [1.6.0] - 2017-06-17

### Fix
- Unable to install (runtimePath)

## [1.5.3] - 2017-06-17

### Dev
- Improve filter on sid, gid and qid

## [1.5.2] - 2017-03-17

### Fix
- [security] XSS in upload file

### Feature
- Export as xlsx (not as xls)

## [1.4.0] - 2017-03-02

### Fix
- Language can be updated during survey
- Update to 2.6.4_lts

### Feature
- Leave partial HTML in admin email

## [1.3.1] - 2017-02-02

### Fix
- Hidden Welcome page can be shown with token (Thank to [@Ben_V](http://www.pmidcalc.org/)) (re-fix)

### Translation
- Update catalan translation (Thanks to [@Ben_V](http://www.pmidcalc.org/)

## [1.3.0] - 2017-02-02

### Fix
- Import lsa don't import all token

### Feature
- Allow to update via cli
- Allow to update runtime directory in config (LimeSurvey API 3 compatible only)

### [1.2.2] - 2017-01-12
- Update [PHPMailer](https://github.com/PHPMailer/PHPMailer/releases/tag/v5.2.22) for security issue

## [1.2.1] - 2016-12-30

### Fix
- Update [PHPMailer](https://github.com/PHPMailer/PHPMailer/releases/tag/v5.2.21) for security issue

## [1.2.0] - 2016-11-16

### Fix
- Issue #11996: Possible remote code execution [LouisGac](https://github.com/LouisGac)
- Possible session fixation on survey entry with token [Carsten Schmitz](https://github.com/c-schmitz)

## [1.1.2] - 2016-11-16

### Feature
- Allow to import broken lsq and lsg
- Show (all) errors when try to import VV file
- Better gitignore for plugins
- Order survey listing by date desc by default

## [1.1.1] - 2016-11-16

### Fix
- Expression manager convert_value does not obey strict setting (Olle Haerstedt) (2.6.1lts)
- Date/Time filed does not record the answer (Olle Haerstedt) (2.6.1lts)

## [1.1.0] - 2016-11-10

### Fix
- Table in reponse view and response edit can take a big width and are unusuable
- readonly for readonly attribute
- Set secure cookies by default is connexion is secure

### Feature
- Allow self signed or invalid TLS/SSL smtp server (with ssl_allow_self_signed in config)

## [1.0.28] - 2016-10-23

### Fix
- Rank show rtl on ltr
- #11821: Apache error uploading images kcfinder ([Carsten Schmitz](http://limesurvey.org))

## [1.0.27] - 2016-10-06

### Fix
- PHP7 compatibility fix ([SamMousa](http://befound.nl))

## [1.0.26] - 2016-09-29

### Fix
- Statistics show more rank than available
- Max sub-question attribute for ranking question still editable after activation

## [1.0.25] - 2016-09-27

### Fix
- #11455: MapQuest discontinued the free tileserver ([Markus Flür](http://limesurvey.org))

## [1.0.24] - 2016-09-16

### Feature
- Allow plugin to add new question attributes

### Fix
- If no plugin add attribute : no questions are shown
- #10226: {ANSWERTABLE} includes <script> placed in question text ([LouisGac](http://limesurvey.org))
- ~ and _ in tokens hard to manually enter
- #11509: numerical input option integer only leads to positive integer input only

### Updated
- improve Plain text ANSWERTABLE for email
- better ANSWERTABLE for multiple question.

## [1.0.23] - 2016-07-17

### Feature
- Allow plugin to set more option when using renderHtml

### Fix
- Language is not correct when session timed out  or other error page
- Allow plugin to allow preview question or group without admin rights

## [1.0.22] - 2016-07-05

### Feature
- Log error and warning in tmp/runtime/application.log by default
- Allow plugin to set more option when using renderHtml

### Fix
- Improvement on error when DB save for public survey
- Fix DECIMAL value before try to save

## [1.0.21] - 2016-06-17

### Feature
- Allow to set column number on ranking question

### Fix
- EM tips are empty after reloading page via browser (F5)
- Ranking question : Alert are not show every time
- Ranking question : Add answers can broke Survey DB

## [1.0.20] - 2016-06-16

### Feature
- afterSurveyMenuLoad event to add survey specific menu items

### Fix
- Improvement on Conditions page with array (number) questions
- Google Analytics code not running

## [1.0.19] - 2016-06-09

### Fix
- Better loading of plugins for command
- Default email format in survey to html

## [1.0.17] - 2016-05-19

### Feature
- Add cssclass question attribute

### Fix
- Plugin survey setting type "checkbox" does not properly save
- Map question : google.maps fix for hidden text element
- Numeric comparaison with Expression
- Survey response marked as 'finished' after opening email link/password twice
- Unable to export result as PDF
- PHP memory_limit being set too low
- Bad link for Browse uploaded ressources if publicurl is set

### Updated
- New token table firstname/lastname to 150

## [1.0.9] - 2016-04-22

### Feature
- beforeController event plugin (for web)
- newUnsecureRequest : plugin direct request without CRSF

### Fix
- Checked responses are not read when load "surveys uploaded file"
- Using mktime() EM function broke survey administration
- EM regexMatch function don't show pattern error
- Attachments for registration emails don't get attached
- Remote control list_surveys can list whole surveys, and not only needed
- event beforeTokenEmail is not dispatched for register
- thousand separator break slider in some condition

## [1.0.6] - 2016-03-20

### Fix
- Issue with relevance on X scale with multilingual
- Languages can not be updated in label set administration
- Error with SMTP mail method
- Broken HTML or script can broke Survey Logic File
- beforeHasPermission event don't happen for owner of survey
- [Security] Survey ID not properly sanitized on survey creation
- 4-byte UTF characters (e.g. Emojis) entered into text can cause DB issue (mysql)
- [Security] issue when saving/loading responses on public survey

## [1.0.0] - 2016-03-01

### Fix
- Higher risk that the emails are rated as Spam
- Filter script in Plugin management and Survey Logic file.

### Updated

common.php
<?php
/*
 * LimeSurvey
 * Copyright (C) 2007 The LimeSurvey Project Team / Carsten Schmitz
 * All rights reserved.
 * License: GNU/GPL License v2 or later, see LICENSE.php
 * LimeSurvey is free software. This version may have been modified pursuant
 * to the GNU General Public License, and as distributed it includes or
 * is derivative of works licensed under the GNU General Public License or
 * other free or open source software licenses.
 * See COPYRIGHT.php for copyright notices and details.
 *
 * $Id: common.php 12242 2012-01-27 23:41:13Z c_schmitz $
 */

//Security Checked: POST, GET, SESSION, DB, REQUEST, returnglobal

//Ensure script is not run directly, avoid path disclosure
if (!isset($dbprefix) || isset($_REQUEST['dbprefix'])) {safe_die("Cannot run this script directly");}

// Include version information
require($rootdir.'/version.php');
require($rootdir."/common_functions.php");

// Check for most necessary requirements
// Now check for PHP & db version
// Do not localize/translate this!
$ver = explode( '.', PHP_VERSION );
$ver_num = $ver[0] . $ver[1] . $ver[2];
$dieoutput='';
if ( $ver_num < 500 )
{
    $dieoutput .= 'This script can only be run on PHP version 5.x or later! Your version: '.phpversion().'<br />';
}
if (!function_exists('mb_convert_encoding'))
{
    $dieoutput .= "This script needs the PHP Multibyte String Functions library installed: See <a href='http://docs.limesurvey.org/tiki-index.php?page=Installation+FAQ'>FAQ</a> and <a href='http://de.php.net/manual/en/ref.mbstring.php'>PHP documentation</a><br />";
}
if ($dieoutput!='') die($dieoutput);

if (!isset($debug)) {$debug=0;}  // for some older config.php's

if ($debug>0) {//For debug purposes - switch on in config.php
    @ini_set("display_errors", 1);
    error_reporting(E_ALL);
}

if ($debug>2) {//For debug purposes - switch on in config.php
    error_reporting(E_ALL | E_STRICT);
}

if (ini_get("max_execution_time")<1200) @set_time_limit(1200); // Maximum execution time - works only if safe_mode is off
//@ini_set("memory_limit",$memorylimit); // Set Memory Limit for big surveys

$maildebug='';


// The following function (when called) includes FireBug Lite if true
define('FIREBUG' , $use_firebug_lite);

define('ADODB_ASSOC_CASE', 2); // needed to set proper upper/lower casing for mssql

##################################################################################

require_once ($rootdir.'/classes/adodb/adodb.inc.php');
require_once ($rootdir.'/classes/datetimeconverter/class.datetimeconverter.php');
require_once ($rootdir.'/classes/phpmailer/class.phpmailer.php');
require_once ($rootdir.'/classes/php-gettext/gettext.inc');
require_once ($rootdir.'/classes/core/surveytranslator.php');
require_once ($rootdir.'/classes/core/sanitize.php');

//  DB session handling
if ($sessionhandler=='db')
{
    require_once($rootdir."/classes/adodb/session/adodb-session2.php");
    $sessionoptions['table'] = $dbprefix.'sessions';
    ADOdb_Session::config($databasetype, $databaselocation, $databaseuser, $databasepass, $databasename, $sessionoptions);
}


$dbprefix=strtolower($dbprefix);
define("_PHPVERSION", phpversion()); // This is the same as the server defined 'PHP_VERSION'


// Deal with server systems having not set a default time zone
if(function_exists("date_default_timezone_set") and function_exists("date_default_timezone_get"))
@date_default_timezone_set(@date_default_timezone_get());


//Every 50th time clean up the temp directory of old files (older than 1 day)
//depending on the load the  probability might be set higher or lower
if (rand(1,50)==25)
{
    cleanTempDirectory();
}

// Array of JS and CSS scripts to include in client header
$js_header_includes = array();
$css_header_includes =  array();

// JS scripts and CSS to include in admin header
// updated by admin scripts
$js_admin_includes = array();
$css_admin_includes = array();

/*
 * $sourcefrom variable checks the location of the current script against
 * the administration directory, and if the current script is running
 * in the administration directory, it is set to "admin". Otherwise it is set
 * to "public". When $sourcefrom is "admin" certain administration only functions
 * are loaded.
 */

$scriptlocation=realpath(".");
$slashlesspath=str_replace(array("\\", "/"), "", $scriptlocation);
$slashlesshome=str_replace(array("\\", "/"), "", $homedir);

// Uncomment the following line for debug purposes
// echo $slashlesspath." - ".$slashlesshome;

// Detect command line usage
if (!empty($argc) && basename($argv[0]) == "cmdline_importsurvey.php") {
    $sourcefrom="admin";
} else if (strcasecmp($slashlesshome, $slashlesspath) == 0) {
    if (strcasecmp($slashlesshome."install", $slashlesspath) != 0)
    {
        $sourcefrom="admin";
    }
    else
    {
        $sourcefrom="install";
    }
} else {
    $sourcefrom="public";
}

// Set path for captcha verification.php
if ($sourcefrom == "admin")
{
    $captchapath='../';
}
else
{
    $captchapath=$rooturl.'/';
}

//check if databasename is set in config file
if (!$databasename)
{
    die("Database Information not provided. If you try to install LimeSurvey please refer to the <a href='http://docs.limesurvey.org'>installation docs</a> and/or contact the system administrator of this webpage.");
}

//BEFORE SESSIONCONTOL BECAUSE OF THE CONNECTION
//CACHE DATA
$connect=ADONewConnection($databasetype);
$database_exists = FALSE;
switch ($databasetype)
{
    case "postgres":
    case "mysqli":
    case "mysql": if ($databaseport!="default") {$dbhost="$databaselocation:$databaseport";}
    else {$dbhost=$databaselocation;}
    break;
    case "mssql_n":
	case "mssqlnative":
    case "odbtp": if ($databaseport!="default") {$dbhost="$databaselocation,$databaseport";}
    else {$dbhost=$databaselocation;}
    break;
    case "odbc_mssql": $dbhost="Driver={SQL Server};Server=$databaselocation;Database=".$databasename;
    break;

    default: safe_die("Unknown database type");
}
// Now try connecting to the database
if ($databasepersistent==true)
{
    if (@$connect->PConnect($dbhost, $databaseuser, $databasepass, $databasename))
    {
        $database_exists = TRUE;
    }
    else {
        // If that doesnt work try connection without database-name
        $connect->database = '';
        if (!@$connect->PConnect($dbhost, $databaseuser, $databasepass))
        {
            safe_die("Can't connect to LimeSurvey database. Reason: ".$connect->ErrorMsg());
        }
    }
}
else
{
    if (@$connect->Connect($dbhost, $databaseuser, $databasepass, $databasename))
    {
        $database_exists = TRUE;
    }
    else {
        // If that doesnt work try connection without database-name
        $connect->database = '';
        if (!@$connect->Connect($dbhost, $databaseuser, $databasepass))
        {
            safe_die("Can't connect to LimeSurvey database. Reason: ".$connect->ErrorMsg());
        }
    }
}

// AdoDB seems to be defaulting to ADODB_FETCH_NUM and we want to be sure that the right default mode is set

$connect->SetFetchMode(ADODB_FETCH_ASSOC);

$dbexistsbutempty=($database_exists && !tableExists('surveys'));



if ($databasetype=='mysql' || $databasetype=='mysqli') {
    if ($debug>1) { @$connect->Execute("SET SESSION SQL_MODE='STRICT_ALL_TABLES,ANSI'"); } //for development - use mysql in the strictest mode  //Checked
    $infoarray=$connect->ServerInfo();
    if (version_compare ($infoarray['version'],'4.1','<'))
    {
        safe_die ("<br />Error: You need at least MySQL version 4.1 to run LimeSurvey. Your version:".$infoarray['version']);
    }
    @$connect->Execute("SET CHARACTER SET 'utf8'");  //Checked
    @$connect->Execute("SET NAMES 'utf8'");  //Checked
}

// Setting dateformat for mssql driver. It seems if you don't do that the in- and output format could be different
if ($databasetype=='odbc_mssql' || $databasetype=='odbtp' || $databasetype=='mssql_n' || $databasetype=='mssqlnative') {
    @$connect->Execute('SET DATEFORMAT ymd;');     //Checked
    @$connect->Execute('SET QUOTED_IDENTIFIER ON;');     //Checked
}


//check if database exist and access is not via install script
if (!$database_exists && !$cmd_install && (strcasecmp($slashlesspath,str_replace(array("\\", "/"), "", $homedir."install")) != 0)) {
    die ("<br/>The LimeSurvey database does not exist. Please run the <a href='$homeurl/install/index.php'>install script</a> to create the necessary database.");
}

 // Check if the DB is up to date and access is not via install script
if ($dbexistsbutempty && !$cmd_install && (strcasecmp($slashlesspath,str_replace(array("\\", "/"), "", $homedir."install")) != 0)) {
    die ("<br />The LimeSurvey database does exist but it seems to be empty. Please run the <a href='$homeurl/install/index.php'>install script</a> to create the necessary tables.");
}

// Default global values that should not appear in config-defaults.php
$updateavailable=0;
$updatebuild='';
$updateversion='';
$updatelastcheck='';
$updatekey='';
$updatekeyvaliduntil='';

require ($homedir.'/globalsettings.php');
SSL_mode();// This really should be at the top but for it to utilise getGlobalSetting() it has to be here

$showxquestions = getGlobalSetting('showxquestions');
$showgroupinfo = getGlobalSetting('showgroupinfo');
$showqnumcode = getGlobalSetting('showqnumcode');

if ($sourcefrom == "admin")
{
    require_once($homedir.'/admin_functions.php');
}

// Check if the DB is up to date
If (!$dbexistsbutempty && $sourcefrom=='admin')
{
    $usrow = getGlobalSetting('DBVersion');
    if (floatval($usrow)<$dbversionnumber)
    {
        $action='';
        require_once($rootdir.'/classes/core/language.php');
        $clang = new limesurvey_lang($defaultlang);
        include_once($homedir.'/update/updater.php');
        if(isset($_GET['continue']) && $_GET['continue']==1)
        {
            echo getAdminHeader();
            CheckForDBUpgrades();
            echo "<br /><a href='$homeurl'>".$clang->gT("Back to main menu")."</a></div>";
            updatecheck();
            echo getAdminFooter("http://docs.limesurvey.org", $clang->gT("LimeSurvey online manual"));
        }
        else
        {
            echo getAdminHeader(),
                 '<div class="messagebox">',CheckForDBUpgrades(),'</div>',
                 getAdminFooter("http://docs.limesurvey.org", $clang->gT("LimeSurvey online manual"))
                 ;
        }
        die;
    }

      if (is_dir($homedir."/install") && $debug<2)
       {
        die ("<p style='text-align: center; margin-left: auto; margin-right: auto; width: 500px; margin-top: 50px;'><img src='../images/limecursor-handle.png' /><strong>Congratulations</strong><br /><br />Your installation is now complete. The final step is to remove or rename the LimeSurvey installation directory (admin/install) on your server since it may be a security risk.<br /><br />Once this directory has been removed or renamed you will be able to log in to your new LimeSurvey Installation.<br /><br /><a href='admin.php'>Try again</a></p>");
       }
}

//Admin menus and standards
//IF THIS IS AN ADMIN SCRIPT, RUN THE SESSIONCONTROL SCRIPT
if ($sourcefrom == "admin")
{
    include($homedir."/sessioncontrol.php");
    /**
     * @param string $htmlheader
     * This is the html header text for all administration pages
     *
     */
    $htmlheader = getAdminHeader();
}

//SET LANGUAGE DIRECTORY
if ($sourcefrom == "admin")
{
    $langdir="$publicurl/locale/".$_SESSION['adminlang']."/help";
    $langdirlocal="$rootdir/locale/".$_SESSION['adminlang']."/help";

    if (!is_dir($langdirlocal))  // is_dir only works on local dirs
    {
        $langdir="$publicurl/locale/en/help"; //default to english if there is no matching language dir
    }
}

if ($sourcefrom == "admin" && $buildnumber != "" && $updatecheckperiod>0 && $updatelastcheck<date_shift(date("Y-m-d H:i:s"), "Y-m-d H:i:s", "-".$updatecheckperiod." days"))
{
    updatecheck();
}



//SET LOCAL TIME
if (substr($timeadjust,0,1)!='-' && substr($timeadjust,0,1)!='+') {$timeadjust='+'.$timeadjust;}
if (strpos($timeadjust,'hours')===false && strpos($timeadjust,'minutes')===false && strpos($timeadjust,'days')===false)
{
    $timeadjust=$timeadjust.' hours';
}

// SITE STYLES
$setfont = "<font size='2' face='verdana'>";
$singleborderstyle = "style='border: 1px solid #111111'";



// Closing PHP tag intentionally left out - yes, it is okay
